ocaml/typing/typeopt.ml

```diff
@@ -141,26 +155,46 @@ let classify env ty : classification =
       Addr
   | Tlink _ | Tsubst _ | Tpoly _ | Tfield _ ->
       assert false
-
-let array_type_kind env ty =
+  end
+  | Float64 -> Unboxed_float Pfloat64
+  | Bits32 -> Unboxed_int Pint32
+  | Bits64 -> Unboxed_int Pint64
+  | Word -> Unboxed_int Pnativeint
+  | Void ->
+    raise (Error (loc, Unsupported_sort Void))
+
+let array_type_kind_gen ?elt_sort env loc ty =
```

Comment:
I think optional arguments are almost always a code smell.  It's too easy for callers to miss them.  I would prefer if this were an explicit argument (though I think it's clever to make it an option so you don't have to write `type_sort` everywhere, though a comment noting that it's expensive to pass it `None` would be good - probably in the mli).

This would also have the advantage that it would be clearer in the PR where we are adding calls to `type_sort` and where we are not.  And if there are so many call sites that this leads to an annoying diff to make, possibly that points to this approach being to expensive and a need for us to think harder about adding more sorts in more places.

(I expect when you make this change you'll need to change a bunch of call sites in code that deals with primitives.  Please leave a CR there noting that we can/should fix that by adding more checking of primitives in the front end).

Reply:
Made the suggested change and left CRs at places where we pass `~elt_sort:None`